doc/sql.extensions/README.skip_locked.md

```diff
@@ -0,0 +1,95 @@
+# SKIP LOCKED clause (FB 5.0)
+
+`SKIP LOCKED` clause can be used with `SELECT ... WITH LOCK`, `UPDATE` and `DELETE` statements.
+
+It makes engine skip records locked by others transactions instead of wait on them or raise conflict errors.
+
+This is very useful to implement work queues where one or more processes post work to a table and issue
+an event while workers listen for events and read/delete items from the table. Using `SKIP LOCKED` multiple
+workers can get exclusive work items from the table without conflicts.
+
+## Syntax
+
+```
+SELECT
+  [FIRST ...]
+  [SKIP ...]
+  FROM <sometable>
+  [WHERE ...]
+  [PLAN ...]
+  [ORDER BY ...]
+  [{ ROWS ... } | {OFFSET ...} | {FETCH ...}]
+  [FOR UPDATE [OF ...]]
+  [WITH LOCK [SKIP LOCKED]]
+```
+
+```
+UPDATE <sometable>
+  SET ...
+  [WHERE ...]
+  [PLAN ...]
+  [ORDER BY ...]
+  [ROWS ...]
+  [SKIP LOCKED]
+  [RETURNING ...]
+```
+
+```
+DELETE FROM <sometable>
+  [WHERE ...]
+  [PLAN ...]
+  [ORDER BY ...]
+  [ROWS ...]
+  [SKIP LOCKED]
+  [RETURNING ...]
+```
+
+## Notes
+
+When `SELECT ... WITH LOCK` or `DELETE`/`UPDATE` statements are used without `SKIP LOCKED`
+but with `FIRST`/`SKIP`/`ROWS`/`OFFSET`/`FETCH` clauses (say, "skip/limit" clauses), lock happens
+in the last step (i.e. after the "skip/limit" is done).
+
+When `SKIP LOCKED` is used with "skip/limit", "lock / skip locked" happens before "skip/limit".
```

Comment:
Coud you explain: does it means that `SELECT SKIP 100 ... FROM ... WITH LOCK SKIP LOCKED` will lock 100 skipped records ?

Reply:
Vlas has raised a good point - only rows returned from this query must be locked, this is by design of the WITH LOCK feature. So SKIP clause should skip records before locking. Maybe correct order should be SKIP->LOCK->FIRST? AFAIU, it should not break anything for WITH LOCK, as I don't see how swapping FIRST/LOCK would cause any difference. But for WITH LOCK SKIP LOCKED this would work more correctly than now. The only remaining question is SINGULAR, as it should be the last but it always fetches one more row, thus locking it. But it will throw an error is the second record exists, which is likely to cause a rollback and release the redundant lock.

Reply:
I set this behavior based on tests of Oracle and PgSQL but saved no tests.
I will do it again and post here.

Reply:
# PostgreSQL:

It locks skipped records.

## Prepare:
```
create table t(n integer);

insert into t values (0);
insert into t values (1);
insert into t values (2);
insert into t values (3);
insert into t values (4);
insert into t values (5);
insert into t values (6);
insert into t values (7);
insert into t values (8);
insert into t values (9);
```

## Session 1:
```
postgres=# begin transaction;
BEGIN
postgres=*# 
postgres=*# select * from t order by n offset 3 limit 2 for update;
 n 
---
 3
 4
(2 rows)

```

## Session 2:
```
postgres=# select * from t order by n offset 2 limit 2 for update skip locked;
 n 
---
 7
 8
(2 rows)
```

Reply:
I set this behavior because with `WITH LOCK` (without `SKIP LOCKED`) it "skips, limit and lock" - in this order.

This behavior is incompatible with `SKIP LOCKED` - it does not make sense to limit before lock/skip.

Also made it because PostgreSQL work in this way.

But when changing this it created different behavior for skip with `WITH LOCK` and `WITH LOCK SKIP LOCKED`. This is not good.

I see no good case for SKIP with `WITH LOCKED SKIP LOCKED` (as well with `WITH LOCK`).

So I may revert this behavior.

What I insist is to LOCK/SKIP before limit (FIRST / ROWS) when `SKIP LOCKED` is used.

We may even prohibit SKIP with `WITH LOCK SKIP LOCKED`.

Now for the case of "skip, limit and lock" vs "skip, lock, limit" order, the later should be problematic without `SKIP LOCKED`, so there should be this difference anyway.

Reply:
> Maybe correct order should be SKIP->LOCK->FIRST? AFAIU, it should not break anything for WITH LOCK, as I don't see how swapping FIRST/LOCK would cause any difference.

AFAIK it would work wrongly, locking one more row than ROWS clause wants.

Reply:
> > Maybe correct order should be SKIP->LOCK->FIRST? AFAIU, it should not break anything for WITH LOCK, as I don't see how swapping FIRST/LOCK would cause any difference.
> 
> AFAIK it would work wrongly, locking one more row than ROWS clause wants.

Looks like you are correct, it should work as FirstRowsStream verify it before call the next stream.

Reply:
> The only remaining question is SINGULAR

AFAIU `WITH LOCK` is not allowed in this context.

Reply:
As for Postgres, this is explicitly documented:
> If a LIMIT is used, locking stops once enough rows have been returned to satisfy the limit (but note that **rows skipped over by OFFSET will get locked**)
However, this behaviour really looks weird. It's like `UPDATE ... ROWS 5 TO 10` updating ten records instead of five. But PG does not have any LIMIT-like clause for `UPDATE/DELETE`, so this may be not so counter-intuitive for them as for us.
MariaDB does not document their behaviour for co-existing `OFFSET` and `FOR UPDATE`. Their `UPDATE` supports only `LIMIT` clause (same as our `FIRST`), without the `OFFSET` part. So again this does not shows them any inconsistencies even if working like PG.
Personally, I don't see any practical sense in `SKIP/OFFSET` used together with `SKIP LOCKED`, be it inside `SELECT` or `UPDATE/DELETE`. So I won't be objecting to prohibit this. But if we're able to make it working consistently (skip -> lock/skip locked -> first), this is IMHO also acceptable.

Reply:
> > The only remaining question is SINGULAR
> 
> AFAIU `WITH LOCK` is not allowed in this context.

Yes, it looks so. Good ;-)